### ISSUE TITLE
Deadlock closing app after failed tile request

### ISSUE THREAD

**Issue by springmeyer** [mapbox/mapbox-gl-native#1477]
To replicate:

  - launch the app, hit `s` until you get to the mapbox-streets style.
  - edit `./style/style/mapbox-streets-v7.json` and give a phony url for one of the layers:

```diff
diff --git a/styles/mapbox-streets-v7.json b/styles/mapbox-streets-v7.json
index 565acd2..d83e0fc 100644
--- a/styles/mapbox-streets-v7.json
+++ b/styles/mapbox-streets-v7.json
@@ -168,8 +168,8 @@
   },
   "sources": {
     "mapbox://mapbox.mapbox-streets-v6": {
-      "url": "mapbox://mapbox.mapbox-streets-v6",
-      "type": "vector"
+      "type": "vector",
+      "url":"http://something.thatdoesnotexist.net"
     },
     "mapbox://mapbox.mapbox-terrain-v2": {
       "url": "mapbox://mapbox.mapbox-terrain-v2",
```

 - launch the app and just the terrain will load
 - then try to close the app and it will not close since it will be hung in a deadlock

Seen with `mac osx` and https://github.com/mapbox/mapbox-gl-native/commit/21c419bd91fcb6fc518a6fcf0f8d99a44b673fe6. The Activity Monitor callstack looks like: https://gist.github.com/springmeyer/b39debea42daa9f3e472

**Comment by kkaefer** [comment]
Looks like the request never gets canceled. This is a regression of https://github.com/mapbox/mapbox-gl-native/commit/9781785ab73e8394e8b92625cc4741952f47955d which was removed again in https://github.com/mapbox/mapbox-gl-native/issues/1336

**Comment by kkaefer** [comment]
This regressed as part of #1065; we are no longer sending an explicit termination message to the Map thread, so we never destruct ~MapContext, never destruct ~Source, never cancel the request.

**Comment by jfirebaugh** [comment]
`~MapContext` is called when the [Thread object](https://github.com/mapbox/mapbox-gl-native/blob/5c7867f435f9d94109ce11fcbbd153b016e02fe0/src/mbgl/util/thread.hpp#L138-138) goes out of scope. Something else must be going on, I'll investigate.

**Comment by jfirebaugh** [comment]
Ah, I think I see what you mean about an explicit termination message @kkaefer. I think that should be avoidable with the current design though, where destructors get called through the course of shutting down the thread.

The problem in this scenario is that MapContext's run loop doesn't exit because there are referenced uv_handles belonging to [these objects](https://github.com/mapbox/mapbox-gl-native/blob/bb788f56f268cfebba2105be7e23744779cad7b8/platform/darwin/http_request_nsurl.mm#L72-73) and [this async](https://github.com/mapbox/mapbox-gl-native/blob/19881f8b24c94ef66cc3ee72eebc73eeb37af2af/include/mbgl/storage/request.hpp#L40-40). If I cause these handles to be unref'd, the deadlock is resolved. However then I hit an assertion failure in `~HTTPContext()`. Continuing to investigate.

**Comment by jfirebaugh** [comment]
The assertion failure is because, even with unref'd handles, `~Source` is not getting called. This is due to `Source::load`'s use of `shared_from_this()`, the result of which is bound in the lambda. This forms a reference loop: the lambda holds a reference to `shared_ptr<Source>`, preventing `~Source`, preventing the request from being cancelled, preventing the lambda from being released. Fortunately `shared_from_this()` is unnecessary.

**Comment by jfirebaugh** [comment]
Further untangling is running into the known issue that [cache requests can outlive their completion callback](https://github.com/mapbox/mapbox-gl-native/pull/1369#issuecomment-100320724) and also the complexities of the `DefaultFileSource::cancel` flow. I think I'll need to fix both of these issues in order to fully resolve this, but I know that @kkaefer was also working on cancellation. @kkaefer, how is that looking?

**Comment by springmeyer** [comment]
@jfirebaugh curious, is #1629 a step towards resolving this or unrelated?

**Comment by jfirebaugh** [comment]
I thought it would be a step towards resolving this but in the end it's unrelated.

**Comment by tmpsantos** [comment]
#1588 and #1632 are adding an error notification to renderStill(). We could (and probably should) use the some code to send error notifications on the continuous mode.

**Comment by mikemorris** [comment]
I think I'm hitting a variation of this at https://github.com/mapbox/mapbox-gl-native/pull/1588#issuecomment-106013260

**Comment by tmpsantos** [comment]
@jfirebaugh do you mind if I take this one? I'm on https://github.com/mapbox/mapbox-gl-native/pull/1657 which is almost the same thing (if not the same).

**Comment by jfirebaugh** [comment]
Go for it. I think #1657 is the same. [This async handle](https://github.com/mapbox/mapbox-gl-native/blob/19881f8b24c94ef66cc3ee72eebc73eeb37af2af/include/mbgl/storage/request.hpp#L40-40) keeps the map thread event loop alive.

**Comment by tmpsantos** [comment]
Fixed by #1657
